README.md
Hades
=====
Hades is the AG DSN authentication and authorization system

Documentation for Hades can be found on [here](https://agdsn.github.io/hades/)


Building
========
Enter the `hades` docker container:
```shell
docker compose exec --user=builder --workdir=/build/hades hades bash
````

<details><summary>Building arpreq</summary>

```shell
(
   cd vendor \
   && (cd arpreq && dpkg-buildpackage --no-sign -b) \
   && sudo dpkg -i python3-arpreq_*.deb;
)
```
</details>

<details><summary>Building hades</summary>

```shell
dpkg-buildpackage --no-sign -b \
&& sudo dpkg -i ../hades_*.deb \
&& sudo systemctl start hades
```
</details>

Troubleshooting
===============
<details><summary>Issues with the db schema / access</summary>
Try recreating the costgresql-cluster in the container:

```shell
sudo systemctl stop hades-database
sudo /usr/lib/hades/control-database.sh clear
sudo systemctl start hades-database
```
</details>

<details><summary><code>setup.py</code> cannot be deleted when rebuilding</summary>
This can happen when the template instantiation command fails,
which causes <code>setup.py</code> to not be emitted.
Thus, <code>setup.py clean</code> cannot possibly work.
The quickfix is to

```shell
touch /build/hades/setup.py
```
and then rebuild.
</details>


debian/changelog
hades (0.5.1) bullseye; urgency=high

  * Various important bugfixes in the RPC api (hades-agent):
    * Fix bug breaking any RPC method call (GH-119)
    * Fix bug breaking the `get_auth_attempts_at_port` RPC method
    * Fix two bugs breaking the `get_auth_dhcp_leases_of_ip` RPC method
    * Fix bug breaking the `get_dhcp_leases_of_mac` RPC method

 --  <lukas.juhrich@agdsn.de>  Tue, 20 Sep 2022 12:28:00 +0000

hades (0.5.0) bullseye; urgency=low

  * Upgrade to Debian bullseye (GH-94, GH-99, GH-103, GH-113)
  * Dropped support for Python < 3.8
  * Forceful DHCP lease release (GH-67)
  * Maintain DHCP leases in database using dnsmasq lease script
  * AMQP changes
    * Separate routing keys for each VRRP instance
    * Durable exchanges
    * Prefix option for AMQP objects
    * New default name
  * Improved documentation (GH-73, GH-97)
  * Improved type annotations
  * Build improvements
    * Debian packaging (GH-96)
    * Configure script (GH-100)
  * systemd improvements (GH-102, GH-108)
    * Rename radius-vrrp to root-vrrp for consistency (all other VRRP
      instances are named like their namespace)
    * Order files consistently
    * Use Type=notify for unbound
    * Set RefuseManualStart for netns units, because they would automatically
      if not required anyway
    * Convert the hades target into service, so that it can be reloaded
      Fix reference to non-existent hades-auth-dns.service unit
    * Stop automatic restarts if non-transient setup errors occurr, that
      require operator intervention.
  * Allow authentication with unknown MAC addresses (GH-90)
  * Allow same MAC address in different IP subnets (GH-92)

 -- Sebastian Schrader <sebastian.schrader@agdsn.de>  Sun, 11 Sep 2022 19:04:09 +0000

hades (0.4.0) stretch; urgency=low

  * Start 0.4.0 branch
  * Protect config files with stricter permissions (See GH-69)
  * Manage networking with systemd and ifupdown
  * Improve unit file dependencies
  * Replace leftover static paths in unit files with configured paths
  * Remove ip-netns calls in units
  * Serve static files with nginx (See GH-56)
  * Add dependency for ssl-cert package
  * Mode and group of generated files can be specified via --mode and --group
  * Improve reporting of config related errors (See GH-12)
  * Support relative imports in config files
  * Support multiple template directories, that can override each other
  * Fix broken network_ip option check (GH-77)
  * Add EAP-MD5 support for D-Links DGS-3100
  * Add signed Celery messages

 -- Sebastian Schrader <sebastian.schrader@agdsn.de>  Fri, 25 Aug 2017 23:11:12 +0200

hades (0.3.2-1) stretch; urgency=low

  * New minor release
  * Various bug fixes
  * Working Celery RPC API

 -- Sebastian Schrader <sebastian.schrader@agdsn.de>  Sat, 05 Aug 2017 05:32:05 +0000

hades (0.3.1-1) stretch; urgency=low

  * New minor release

 -- Stefan Haller <stefan.haller@agdsn.de>  Sat, 10 Jun 2017 21:32:38 +0200

hades (0.3.0-1) stretch; urgency=low

  * Initial package release

 -- Sebastian Schrader <sebastian.schrader@agdsn.de>  Sat, 20 May 2017 02:56:47 +0200


conf/config.py
from collections import OrderedDict
import netaddr

HADES_SITE_NAME = 'dev'
HADES_SITE_NODE_ID = 'dev-01'
BROKER_URL = 'amqp://guest:guest@127.0.0.1/hades'
HADES_CREATE_DUMMY_INTERFACES = True
HADES_BRIDGE_SERVICE_INTERFACES = True
HADES_LOCAL_MASTER_DATABASE = True
HADES_AUTH_INTERFACE = 'auth'
HADES_UNAUTH_INTERFACE = 'unauth'
HADES_VRRP_INTERFACE = 'vrrp'
HADES_VRRP_LISTEN_AUTH = netaddr.IPNetwork('192.168.32.2/24')
HADES_VRRP_LISTEN_ROOT = netaddr.IPNetwork('192.168.32.1/24')
HADES_VRRP_LISTEN_UNAUTH = netaddr.IPNetwork('192.168.32.3/24')
HADES_VRRP_PASSWORD = 'correcthorsebatterystaple'
HADES_RADIUS_LISTEN = (
    netaddr.IPNetwork('10.10.10.240/24'),
)
HADES_RADIUS_INTERFACE = 'radius'
HADES_RADIUS_LOCALHOST_SECRET = 'testing123'
HADES_CONTACT_ADDRESSES = OrderedDict([
    ("Support", "support@example.com"),
])
HADES_USER_NETWORKS = {
    'Wu9': netaddr.IPNetwork('141.30.202.1/24'),
    'Wu3': netaddr.IPNetwork('141.30.223.1/24'),
    'Wu1': netaddr.IPNetwork('141.30.224.1/24'),
    'Wu11': netaddr.IPNetwork('141.30.216.1/24'),
    'Wu7': netaddr.IPNetwork('141.30.222.1/24'),
    'Zw41': netaddr.IPNetwork('141.30.226.1/23'),
    'Wu5': netaddr.IPNetwork('141.30.228.1/24'),
    'UNEP': netaddr.IPNetwork('141.30.242.129/28'),
    'Bor34': netaddr.IPNetwork('141.76.121.1/24'),
    'Ger38': netaddr.IPNetwork('141.76.124.1/24'),
    'Bu22': netaddr.IPNetwork('141.30.204.1/24'),
    'Bu24': netaddr.IPNetwork('141.30.205.1/24'),
    'HSS46a': netaddr.IPNetwork('141.30.217.1/24'),
    'HSS46b': netaddr.IPNetwork('141.30.234.1/25'),
    'HSS48a': netaddr.IPNetwork('141.30.218.1/24'),
    'HSS48b': netaddr.IPNetwork('141.30.215.129/25'),
    'HSS50': netaddr.IPNetwork('141.30.219.1/24'),
    'Zeu1fa': netaddr.IPNetwork('141.30.234.129/26'),
    'Zeu1fb': netaddr.IPNetwork('141.30.234.193/27 '),
}
HADES_POSTGRESQL_USER_MAPPINGS = {
    'hades-agent': {
        'user': 'fdw',
        'password': 'correcthorsebatterystaple',
    },
    'hades-database': {
        'user': 'fdw',
        'password': 'correcthorsebatterystaple',
    },
    'hades-portal': {
        'user': 'fdw',
        'password': 'correcthorsebatterystaple',
    },
    'hades-radius': {
        'user': 'fdw',
        'password': 'correcthorsebatterystaple',
    },
}
HADES_UNAUTH_WHITELIST_DNS = ('agdsn.de',)
HADES_AUTH_DNS_ALTERNATIVE_ZONES = {
    'example.com': {
        'type': 'transparent',
        'records': [
            {
                'name': 'www.example.com',
                'ttl': 60*60,
                'class': 'IN',
                'type': 'A',
                'data': '127.0.0.1'
            },
        ]
    }
}
HADES_CUSTOM_IPTABLES_INPUT_RULES = [
    # Expose PostgreSQL
    "-p tcp -m tcp --dport 5432 -j ACCEPT",
    # Expose RabbitMQ AMQP socket
    "-p tcp -m tcp --dport 5672 -j ACCEPT",
    # Expose RabbitMQ HTTP management interface
    "-p tcp -m tcp --dport 15672 -j ACCEPT",
]
HADES_POSTGRESQL_LISTEN = [netaddr.IPNetwork("0.0.0.0/0")]
